File: lib/core/utils/formatters.dart
import 'package:intl/intl.dart';

/// Centralized formatting utility for currency, dates, numbers, and more
class Formatters {
  // ==================== CURRENCY ====================

  /// Format amount as currency with symbol (default: $)
  /// Example: formatCurrency(1234.5) → "$1,234.50"
  static String formatCurrency(double amount, {String symbol = '\$'}) {
    final formatted = NumberFormat('#,##0.00').format(amount);
    return '$symbol$formatted';
  }

  /// Format amount with symbol suffix (e.g., "1,234.50 USD")
  /// Example: formatAmountWithSymbol(1234.5, 'USD') → "1,234.50 USD"
  static String formatAmountWithSymbol(double amount, String currencyCode) {
    final formatted = NumberFormat('#,##0.00').format(amount);
    return '$formatted $currencyCode';
  }

  /// Legacy alias for currency (deprecated, use formatCurrency)
  static String currency(double amount, {String symbol = '\$'}) {
    return formatCurrency(amount, symbol: symbol);
  }

  // ==================== DATES ====================

  /// Format date as readable string (e.g., "Jan 15, 2025")
  /// Example: formatDate(DateTime(2025, 1, 15)) → "Jan 15, 2025"
  static String formatDate(DateTime date) {
    return DateFormat('MMM dd, yyyy').format(date);
  }

  /// Format date and time (e.g., "Jan 15, 2025 2:30 PM")
  /// Example: formatDateTime(DateTime(2025, 1, 15, 14, 30)) → "Jan 15, 2025 2:30 PM"
  static String formatDateTime(DateTime dateTime) {
    return DateFormat('MMM dd, yyyy h:mm a').format(dateTime);
  }

  /// Format as ISO date (YYYY-MM-DD)
  /// Example: formatDateISO(DateTime(2025, 1, 15)) → "2025-01-15"
  static String formatDateISO(DateTime date) {
    return DateFormat('yyyy-MM-dd').format(date);
  }

  /// Format as time only (e.g., "2:30 PM")
  /// Example: formatTime(DateTime(2025, 1, 15, 14, 30)) → "2:30 PM"
  static String formatTime(DateTime dateTime) {
    return DateFormat('h:mm a').format(dateTime);
  }

  /// Legacy alias for dateTime (deprecated, use formatDateTime)
  static String date(DateTime date) {
    return formatDate(date);
  }

  /// Legacy alias for dateTime (deprecated, use formatDateTime)
  static String dateTime(DateTime dateTime) {
    return formatDateTime(dateTime);
  }

  // ==================== NUMBERS ====================

  /// Format integer with thousand separators
  /// Example: formatNumber(1234567) → "1,234,567"
  static String formatNumber(int number) {
    return NumberFormat('#,###').format(number);
  }

  /// Format decimal number with specified decimal places
  /// Example: formatDecimal(1234.5, 2) → "1,234.50"
  static String formatDecimal(double number, {int decimals = 2}) {
    final pattern = '#,##0.${'0' * decimals}';
    return NumberFormat(pattern).format(number);
  }

  // ==================== PERCENTAGES ====================

  /// Format as percentage (0.25 → "25.0%")
  /// Example: formatPercentage(0.25) → "25.0%"
  static String formatPercentage(double value, {int decimals = 1}) {
    final percent = value * 100;
    return '${percent.toStringAsFixed(decimals)}%';
  }

  /// Legacy alias (deprecated, use formatPercentage)
  static String percentage(double value) {
    return formatPercentage(value);
  }

  // ==================== INVOICE ====================

  /// Format invoice number with prefix and padded number
  /// Example: formatInvoiceNumber('INV-', 42) → "INV-0042"
  static String formatInvoiceNumber(String prefix, int number, {int padding = 4}) {
    final padded = number.toString().padLeft(padding, '0');
    return '$prefix$padded';
  }

  // ==================== PHONE ====================

  /// Format phone number (10-digit US format)
  /// Example: formatPhone('1234567890') → "(123) 456-7890"
  static String formatPhone(String phone) {
    if (phone.length == 10) {
      return '(${phone.substring(0, 3)}) ${phone.substring(3, 6)}-${phone.substring(6)}';
    }
    return phone;
  }

  /// Legacy alias (deprecated, use formatPhone)
  static String phone(String phone) {
    return formatPhone(phone);
  }

  // ==================== UTILITY HELPERS ====================

  /// Check if a value is a valid number
  static bool isValidNumber(String value) {
    return double.tryParse(value) != null;
  }

  /// Get currency symbol for a currency code
  static String getCurrencySymbol(String currencyCode) {
    final symbols = {
      'USD': '\$',
      'EUR': '€',
      'GBP': '£',
      'JPY': '¥',
      'INR': '₹',
      'BRL': 'R\$',
      'CAD': 'C\$',
      'AUD': 'A\$',
      'CHF': 'CHF',
      'CNY': '¥',
      'SEK': 'kr',
      'NZD': 'NZ\$',
      'MXN': '\$',
      'SGD': 'S\$',
      'HKD': 'HK\$',
    };
    return symbols[currencyCode] ?? currencyCode;
  }
}

File: functions/src/utils/formatters.ts
/**
 * Centralized formatting utilities for Cloud Functions
 * Mirrors Dart formatters for consistency across platforms
 */

// ==================== CURRENCY ====================

/**
 * Format amount as currency with symbol (default: $)
 * @param amount - The amount to format
 * @param symbol - Currency symbol (default: "$")
 * @returns Formatted currency string (e.g., "$1,234.50")
 */
export function formatCurrency(amount: number, symbol: string = '$'): string {
  const formatted = amount.toLocaleString('en-US', {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  });
  return `${symbol}${formatted}`;
}

/**
 * Format amount with symbol suffix (e.g., "1,234.50 USD")
 * @param amount - The amount to format
 * @param currencyCode - Currency code (e.g., "USD")
 * @returns Formatted amount string (e.g., "1,234.50 USD")
 */
export function formatAmountWithSymbol(amount: number, currencyCode: string): string {
  const formatted = amount.toLocaleString('en-US', {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  });
  return `${formatted} ${currencyCode}`;
}

// ==================== DATES ====================

/**
 * Format date as readable string (e.g., "Jan 15, 2025")
 * @param date - Date to format
 * @returns Formatted date string
 */
export function formatDate(date: Date | string): string {
  const d = typeof date === 'string' ? new Date(date) : date;
  return d.toLocaleDateString('en-US', {
    month: 'short',
    day: '2-digit',
    year: 'numeric',
  });
}

/**
 * Format date and time (e.g., "Jan 15, 2025 2:30 PM")
 * @param dateTime - DateTime to format
 * @returns Formatted datetime string
 */
export function formatDateTime(dateTime: Date | string): string {
  const dt = typeof dateTime === 'string' ? new Date(dateTime) : dateTime;
  const date = dt.toLocaleDateString('en-US', {
    month: 'short',
    day: '2-digit',
    year: 'numeric',
  });
  const time = dt.toLocaleTimeString('en-US', {
    hour: 'numeric',
    minute: '2-digit',
    hour12: true,
  });
  return `${date} ${time}`;
}

/**
 * Format as ISO date (YYYY-MM-DD)
 * @param date - Date to format
 * @returns ISO date string
 */
export function formatDateISO(date: Date | string): string {
  const d = typeof date === 'string' ? new Date(date) : date;
  return d.toISOString().split('T')[0];
}

/**
 * Format as time only (e.g., "2:30 PM")
 * @param dateTime - DateTime to format
 * @returns Formatted time string
 */
export function formatTime(dateTime: Date | string): string {
  const dt = typeof dateTime === 'string' ? new Date(dateTime) : dateTime;
  return dt.toLocaleTimeString('en-US', {
    hour: 'numeric',
    minute: '2-digit',
    hour12: true,
  });
}

// ==================== NUMBERS ====================

/**
 * Format integer with thousand separators
 * @param number - Number to format
 * @returns Formatted number string (e.g., "1,234,567")
 */
export function formatNumber(number: number): string {
  return Math.floor(number).toLocaleString('en-US');
}

/**
 * Format decimal number with specified decimal places
 * @param number - Number to format
 * @param decimals - Number of decimal places (default: 2)
 * @returns Formatted decimal string
 */
export function formatDecimal(number: number, decimals: number = 2): string {
  return number.toLocaleString('en-US', {
    minimumFractionDigits: decimals,
    maximumFractionDigits: decimals,
  });
}

// ==================== PERCENTAGES ====================

/**
 * Format as percentage (0.25 → "25.0%")
 * @param value - Value as decimal (0-1)
 * @param decimals - Number of decimal places (default: 1)
 * @returns Formatted percentage string
 */
export function formatPercentage(value: number, decimals: number = 1): string {
  const percent = value * 100;
  return `${percent.toFixed(decimals)}%`;
}

// ==================== INVOICE ====================

/**
 * Format invoice number with prefix and padded number
 * @param prefix - Invoice prefix (e.g., "INV-")
 * @param number - Invoice number
 * @param padding - Number of digits to pad to (default: 4)
 * @returns Formatted invoice number (e.g., "INV-0042")
 */
export function formatInvoiceNumber(
  prefix: string,
  number: number,
  padding: number = 4
): string {
  const padded = number.toString().padStart(padding, '0');
  return `${prefix}${padded}`;
}

// ==================== PHONE ====================

/**
 * Format phone number (10-digit US format)
 * @param phone - Phone number string
 * @returns Formatted phone string (e.g., "(123) 456-7890")
 */
export function formatPhone(phone: string): string {
  const cleaned = phone.replace(/\D/g, '');
  if (cleaned.length === 10) {
    return `(${cleaned.substring(0, 3)}) ${cleaned.substring(3, 6)}-${cleaned.substring(6)}`;
  }
  return phone;
}

// ==================== UTILITY HELPERS ====================

/**
 * Check if a value is a valid number
 * @param value - Value to check
 * @returns True if valid number
 */
export function isValidNumber(value: string | number): boolean {
  return !isNaN(Number(value));
}

/**
 * Get currency symbol for a currency code
 * @param currencyCode - Currency code (e.g., "USD")
 * @returns Currency symbol
 */
export function getCurrencySymbol(currencyCode: string): string {
  const symbols: { [key: string]: string } = {
    USD: '$',
    EUR: '€',
    GBP: '£',
    JPY: '¥',
    INR: '₹',
    BRL: 'R$',
    CAD: 'C$',
    AUD: 'A$',
    CHF: 'CHF',
    CNY: '¥',
    SEK: 'kr',
    NZD: 'NZ$',
    MXN: '$',
    SGD: 'S$',
    HKD: 'HK$',
  };
  return symbols[currencyCode] || currencyCode;
}

/**
 * Format currency amount with locale awareness
 * @param amount - Amount to format
 * @param currencyCode - ISO currency code
 * @param locale - Locale string (default: "en-US")
 * @returns Formatted currency
 */
export function formatCurrencyLocale(
  amount: number,
  currencyCode: string,
  locale: string = 'en-US'
): string {
  try {
    return new Intl.NumberFormat(locale, {
      style: 'currency',
      currency: currencyCode,
    }).format(amount);
  } catch (e) {
    // Fallback if locale not supported
    return formatAmountWithSymbol(amount, currencyCode);
  }
}

/**
 * Escape HTML special characters (safe for invoice rendering)
 * @param text - Text to escape
 * @returns Escaped HTML
 */
export function escapeHtml(text: string): string {
  if (!text) return '';
  const map: { [key: string]: string } = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#039;',
  };
  return text.replace(/[&<>"']/g, (char) => map[char]);
}

/**
 * Escape CSV special characters
 * @param text - Text to escape
 * @returns Escaped CSV value
 */
export function escapeCsv(text: string): string {
  if (!text) return '';
  if (text.includes(',') || text.includes('"') || text.includes('\n')) {
    return `"${text.replace(/"/g, '""')}"`;
  }
  return text;
}
